api/testing/restapi_test.go

```diff
@@ -0,0 +1,686 @@
+	}
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Attach(id, gomock.Any()).
+			Return("", fmt.Errorf("some error")),
+	)
+
+	// create driver client
+
+	driverclient := volumeclient.VolumeDriver(ts.client)
+
+	_, err = driverclient.Attach(id, req.GetOptions())
+
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "some error")
+
+}
+
+func TestVolumeDetachSuccess(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+
+	name := "myvol"
+	id := "myid"
+	size := uint64(10)
+
+	req := &api.VolumeSetRequest{
+		Options: map[string]string{},
+		Action: &api.VolumeStateAction{
+			Attach: api.VolumeActionParam_VOLUME_ACTION_PARAM_OFF},
+		Locator: &api.VolumeLocator{Name: name},
+		Spec:    &api.VolumeSpec{Size: size},
+	}
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Detach(id, gomock.Any()).
+			Return(nil),
+
+		ts.MockDriver().
+			EXPECT().
+			Inspect([]string{id}).
+			Return([]*api.Volume{
+				&api.Volume{
+					Id: id,
+					Locator: &api.VolumeLocator{
+						Name: name,
+					},
+					Spec: &api.VolumeSpec{
+						Size: size,
+					},
+				},
+			}, nil),
+	)
+
+	// create client
+
+	driverclient := volumeclient.VolumeDriver(ts.client)
+	res := driverclient.Detach(id, req.GetOptions())
+
+	assert.Nil(t, res)
+}
+
+func TestVolumeDetachFailed(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+
+	name := "myvol"
+	id := "myid"
+	size := uint64(10)
+
+	req := &api.VolumeSetRequest{
+		Options: map[string]string{},
+		Action: &api.VolumeStateAction{
+			Attach: api.VolumeActionParam_VOLUME_ACTION_PARAM_OFF},
+		Locator: &api.VolumeLocator{Name: name},
+		Spec:    &api.VolumeSpec{Size: size},
+	}
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Detach(id, gomock.Any()).
+			Return(fmt.Errorf("Error in detaching")),
+	)
+
+	// create client
+
+	driverclient := volumeclient.VolumeDriver(ts.client)
+	res := driverclient.Detach(id, req.GetOptions())
+
+	assert.NotNil(t, res)
+	assert.Contains(t, res.Error(), "Error in detaching")
+}
+
+func TestVolumeMountSuccess(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	var err error
+
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+	assert.Nil(t, err)
+
+	name := "myvol"
+	id := "myid"
+	size := uint64(10)
+
+	//create request
+	req := &api.VolumeSetRequest{
+		Options: map[string]string{},
+		Action: &api.VolumeStateAction{
+			Attach:    api.VolumeActionParam_VOLUME_ACTION_PARAM_ON,
+			Mount:     api.VolumeActionParam_VOLUME_ACTION_PARAM_ON,
+			MountPath: "/mnt"},
+		Locator: &api.VolumeLocator{Name: name},
+		Spec:    &api.VolumeSpec{Size: size},
+	}
+
+	gomock.InOrder(
+
+		ts.MockDriver().
+			EXPECT().
+			Mount(id, gomock.Any(), gomock.Any()).
+			Return(nil),
+
```

Comment:
Also the Locator and Spec is not nil which means d.Set() should also be called and there needs to be an equivalent Expect()
https://github.com/libopenstorage/openstorage/blob/master/api/server/volume.go#L150

Reply:
Client wrapper method makes the call to doVolumeSet (in case of detach , mount and unmount) and dovolumeSetGetResponse( in case of attach)

which creates a VolumeSetRequest instance  with the Action attach and options but locator and spec are nil.
Hence on the server side its nil and no call is made to the Set.